components/local-app/pkg/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,235 @@
+func DownloadManifest(ctx context.Context, baseURL string) (res *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("download manifest from %s/manifest.json: %w", baseURL, err)
+		}
+	}()
+
+	murl, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	originalPath := murl.Path
+	murl.Path = filepath.Join(murl.Path, "manifest.json")
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, murl.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(resp.Status)
+	}
+
+	var mf Manifest
+	err = json.NewDecoder(resp.Body).Decode(&mf)
+	if err != nil {
+		return nil, err
+	}
+	for i := range mf.Binaries {
+		murl.Path = filepath.Join(originalPath, mf.Binaries[i].Filename)
+		mf.Binaries[i].URL = murl.String()
+	}
+
+	return &mf, nil
+}
+
+// NeedsUpdate checks if the current version is outdated
+func NeedsUpdate(current *semver.Version, manifest *Manifest) bool {
+	return manifest.Version.GreaterThan(current)
+}
+
+// ReplaceSelf replaces the current binary with the one from the manifest, no matter the version
+// If there is no matching binary in the manifest, this function returns ErrNoBinaryAvailable.
+func ReplaceSelf(ctx context.Context, manifest *Manifest) error {
+	var binary *Binary
+	for _, b := range manifest.Binaries {
+		if b.OS != runtime.GOOS || b.Arch != runtime.GOARCH {
+			continue
+		}
+
+		binary = &b
+		break
+	}
+	if binary == nil {
+		return ErrNoBinaryAvailable
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, binary.URL, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	dgst, _ := hex.DecodeString(binary.Digest.Hex())
+	return update.Apply(resp.Body, update.Options{
+		Checksum:   dgst,
+		Hash:       crypto.SHA256,
+		TargetMode: 0755,
+	})
+}
+
+var ErrNoBinaryAvailable = errors.New("no binary available for this platform")
+
+// Autoupdate checks if there is a newer version available and updates the binary if so
+// actually updates. This function returns immediately and runs the update in the background.
+// The returned function can be used to wait for the update to finish.
+func Autoupdate(ctx context.Context, cfg *config.Config) func() {
+	if !cfg.Autoupdate {
+		return func() {}
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		gpctx, _ := cfg.GetActiveContext()
+		if gpctx == nil {
+			slog.Debug("no active context - autoupdate disabled")
+			return
+		}
+
+		var err error
+		defer func() {
+			if err != nil {
+				slog.Debug("autoupdate failed", "err", err)
+			}
+		}()
+
+		mfctx, cancel := context.WithTimeout(ctx, 1*time.Second)
+		defer cancel()
+		baseURL := *gpctx.Host.URL
+		baseURL.Path = "/static/bin"
```

Comment:
Might be a good idea to either make a subdirectory for all the `gitpod` CLI-related binaries and manifests or rename the manifest to something CLI-specific, so that it doesn't get mixed up with the local companion and any future binaries we might have.

Reply:
I hope we won't have too many different CLIs in the future, but one to rule them all :D